File: src/sp/it/pl/gui/itemnode/textfield/MoodItemNode.java
package sp.it.pl.gui.itemnode.textfield;

import sp.it.pl.audio.tagging.Metadata;
import sp.it.pl.gui.objects.picker.MoodPicker;
import sp.it.pl.gui.objects.popover.NodePos;
import sp.it.pl.gui.objects.popover.PopOver;
import sp.it.pl.util.access.V;
import sp.it.pl.util.parsing.Parser;
import static sp.it.pl.gui.objects.textfield.autocomplete.AutoCompletion.autoComplete;
import static sp.it.pl.main.AppUtil.APP;
import static sp.it.pl.util.functional.Util.filter;

/**
 * {@link sp.it.pl.gui.itemnode.textfield.TextFieldItemNode} for audio mood tagging values.<br/>
 * It provides two additional functionalities
 * <ul>
 * <li> Auto-completion from set of moods application is aware of
 * <li> Mood picker popup. The position of the picker popup can be customized.
 * </ul>
 */
public class MoodItemNode extends TextFieldItemNode<String> {

	/** The position for the picker to show on. */
	public final V<NodePos> pos = new V<>(NodePos.RIGHT_CENTER);

	public MoodItemNode() {
		super(Parser.DEFAULT.toConverterOf(String.class));
		setEditable(true);
		autoComplete(this, p -> filter(APP.db.getStringPool().getStrings(Metadata.Field.MOOD.name()), t -> APP.db.getAutocompletionFilter().apply(t, p.getUserText())));
	}

	@Override
	void onDialogAction() {
		MoodPicker picker = new MoodPicker();
		PopOver<?> p = new PopOver<>(picker.getNode());
		p.detachable.set(false);
		p.setArrowSize(0);
		p.setArrowIndent(0);
		p.setCornerRadius(0);
		p.setAutoHide(true);
		p.setAutoFix(true);
		picker.onCancel = p::hide;
		picker.onSelect = mood -> {
			setValue(mood);
			p.hide();
		};
		picker.getNode().setPrefSize(800, 600);
		p.show(this, pos.get());
	}

}

File: src/sp/it/pl/gui/itemnode/textfield/MoodItemNode.kt
package sp.it.pl.gui.itemnode.textfield

import javafx.scene.Node
import sp.it.pl.audio.tagging.Metadata
import sp.it.pl.gui.objects.picker.MoodPicker
import sp.it.pl.gui.objects.popover.NodePos
import sp.it.pl.gui.objects.popover.PopOver
import sp.it.pl.gui.objects.textfield.autocomplete.AutoCompletion.autoComplete
import sp.it.pl.main.AppUtil.APP
import sp.it.pl.util.access.V
import java.util.function.Consumer

/** Text field for audio mood tagging values with a picker and autocompletion. */
class MoodItemNode: TextFieldItemNode<String>({ APP.converter.general.toS(it) }) {

    /** The position for the picker to show on. */
    val pickerPosition = V(NodePos.RIGHT_CENTER)

    init {
        isEditable = true
        autoComplete(this) { p ->
            APP.db.itemUniqueValuesByField[Metadata.Field.MOOD]
                    ?.filter { it.contains(p.userText, true) }
                    .orEmpty()
        }
    }

    override fun onDialogAction() {
        val p = PopOver<Node>().apply {
            detachable.set(false)
            arrowSize = 0.0
            arrowIndent = 0.0
            cornerRadius = 0.0
            isAutoHide = true
            isAutoFix = true
            contentNode = MoodPicker().apply {
                node.setPrefSize(800.0, 600.0)
                onCancel = Runnable { hide() }
                onSelect = Consumer {
                    value = it
                    hide()
                }
            }.node
        }
        p.show(this, pickerPosition.get())
    }

}